File: Extensions/Tessa.Extensions.Default.Client/AbTest/AbExternalFileSource.cs
#nullable enable
using System;
using System.IO;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Tessa.Files;
using Tessa.Platform.Runtime;
using Tessa.Platform.Storage;
using Unity;

namespace Tessa.Extensions.Default.Client.AbTest
{
    public class AbExternalFileSource(IFileCache cache, ISession session, [OptionalDependency] IFileManager? manager = null)
        : FileSource(cache, session, manager)
    {
        #region Base Overrides

        /// <inheritdoc/>
        protected override async ValueTask<IFile> CreateFileCoreAsync(IFileCreationToken token, IFileContent? content = null, CancellationToken cancellationToken = default)
        {
            // здесь возможно исключение, связанное с параметрами метода
            var typedToken = (AbExternalFileCreationToken) token;

            var name = NotEmptyOrThrow(token.Name);
            IFileContent? allocatedContent = null;
            var actualContent = content ?? (allocatedContent = await this.Cache.AllocateAsync(name, cancellationToken: cancellationToken).ConfigureAwait(false));

            try
            {
                // все свойства токена проверяются в конструкторе
                var file = new AbExternalFile(
                    token.ID ?? Guid.NewGuid(),
                    name,
                    token.Size,
                    token.Category,
                    NotNullOrThrow(token.Type),
                    actualContent,
                    source: this,
                    token.Modified,
                    token.ModifiedByID,
                    token.ModifiedByName,
                    token.Created,
                    token.CreatedByID,
                    token.CreatedByName,
                    token.Permissions.Clone(),
                    token.IsLocal,
                    hash: token.Hash,
                    description: typedToken.Description);

                file.Options.SetStorage(StorageHelper.Clone(token.Options));
                file.Info.SetStorage(StorageHelper.Clone(token.Info));
                file.RequestInfo.SetStorage(StorageHelper.Clone(token.RequestInfo));

                allocatedContent = null;
                return file;
            }
            finally
            {
                if (allocatedContent is not null)
                {
                    await allocatedContent.DisposeAsync().ConfigureAwait(false);
                }
            }
        }

        /// <inheritdoc/>
        protected override ValueTask<IFileCreationToken> GetFileCreationTokenCoreAsync(CancellationToken cancellationToken = default)
        {
            var user = this.Session.User;
            var utcNow = DateTime.UtcNow;
            return new(new AbExternalFileCreationToken
            {
                Modified = utcNow,
                ModifiedByID = user.ID,
                ModifiedByName = user.Name,
                Created = utcNow,
                CreatedByID = user.ID,
                CreatedByName = user.Name
            });
        }

        /// <inheritdoc/>
        protected override async ValueTask<IFileContentResponse> GetContentCoreAsync(IFileContentRequest request, CancellationToken cancellationToken = default)
        {
            // здесь возможно исключение, если передан файл от другого источника
            var typedFile = (AbExternalFile) request.Version.File;

            if (request.ProcessContentActionAsync is { } processContentActionAsync)
            {
                await processContentActionAsync(
                    new MemoryStream(Encoding.UTF8.GetBytes(typedFile.Description ?? string.Empty)),
                    cancellationToken).ConfigureAwait(false);
            }

            return new FileContentResponse(request.Version);
        }

        #endregion
    }
}

File: WebClient SDK/src/default/abTest/abExternalFileSource.ts
import { saveAs } from 'file-saver';
import { AbExternalFileCreationToken } from './abExternalFileCreationToken';
import { AbExternalFile } from './abExternalFile';
import { FileSource, IFileCreationToken, IFile, IFileVersion, FileVersion } from 'tessa/files';
import { merge } from 'tessa/platform/storage/storageHelper';
import { Guid, Result, ResultWithInfo } from 'tessa/platform';
import { IStorage } from 'tessa/platform/storage';
import {
  ValidationResult,
  ValidationResultBuilder,
  ValidationResultType
} from 'tessa/platform/validation';

export class AbExternalFileSource extends FileSource {
  protected createFileCore(token: IFileCreationToken): IFile {
    // здесь возможно исключение, связанное с параметрами метода
    if (!(token instanceof AbExternalFileCreationToken)) {
      throw new Error('Unexpected token type.');
    }

    // все свойства токена проверяются в конструкторе
    const file = new AbExternalFile(
      token.id || Guid.newGuid(),
      token.name,
      token.category,
      token.type,
      this,
      token.permissions.clone(),
      token.modified,
      token.modifiedById,
      token.modifiedByName,
      token.created,
      token.createdById,
      token.createdByName,
      token.isLocal,
      null,
      token.description
    );

    if (token.options && Object.keys(token.options).length > 0) {
      merge(token.options, file.options);
    }

    if (token.info && Object.keys(token.info).length > 0) {
      merge(token.info, file.info);
    }

    if (token.requestInfo && Object.keys(token.requestInfo).length > 0) {
      merge(token.requestInfo, file.requestInfo);
    }

    return file;
  }

  protected getFileCreationTokenCore(): IFileCreationToken {
    return new AbExternalFileCreationToken();
  }

  protected async getContentCore(
    fileOrFileVersion: IFile | IFileVersion,
    _info?: IStorage
  ): Promise<ResultWithInfo<File>> {
    const file =
      fileOrFileVersion instanceof FileVersion ? fileOrFileVersion.file : fileOrFileVersion;

    const builder = new ValidationResultBuilder();
    if (!(file instanceof AbExternalFile)) {
      builder.add(ValidationResult.fromText('Unexpected file type.', ValidationResultType.Error));
    }
    const description = (file as AbExternalFile).description;

    return {
      data: new File([description], fileOrFileVersion.name),
      info: {},
      validationResult: builder.build()
    };
  }

  protected async saveContentCore(
    fileOrFileVersion: IFile | IFileVersion,
    info?: IStorage
  ): Promise<Result<boolean>> {
    const result = await this.getContentCore(fileOrFileVersion, info);
    if (!result.data) {
      return {
        data: false,
        validationResult: result.validationResult
      };
    }

    await saveAs(result.data, fileOrFileVersion.name);

    return {
      data: true,
      validationResult: result.validationResult
    };
  }
}
